lua_cjson.c

```diff
@@ -698,8 +700,20 @@ static void json_append_data(lua_State *l, json_config_t *cfg,
         len = lua_array_length(l, cfg, json);
         if (len > 0 || (len == 0 && !cfg->encode_empty_table_as_object))
             json_append_array(l, cfg, current_depth, json, len);
-        else
-            json_append_object(l, cfg, current_depth, json);
+        else {
+            int as_array = 0;
+            if (lua_getmetatable(l, -1)) {
+                luaL_getmetatable(l, CJSON_AS_ARRAY_MT);
```

Comment:
I am sorry but I really don't know what you mean here.

> Maybe we could just store the metatable in our own cjson module table

I can see this. But then how does one retrieve it for comparison in ` json_append_data`? Do you mean to store it as an upvalue like the configuration?

> with a lightuserdata key

I can see some syntactic sugar here and I can implement a `cjson.empty_array` lightuserdata just like `cjson.null`, but I don't see how storing the metatable with a lightuserdata key allows one to retrieve the said metatable for comparison in `json_append_data` either. Except as an uvalue again, but then the lightuserdata key seems irrelevant.

> to avoid polluting the global Lua registry

One could argue the Lua registry is slow and propitious to naming conflicts, but isn't it made for that exact purpose of storing Lua values needed by C modules? In practice, naming conflicts are not so much of a thing (and the name is now prefixed), and however slow the registry is, using it will still be way more efficient than performing the hacks currently necessary to encore empty arrays anyways.

Am I missing something here?